processor/src/main/java/org/mapstruct/ap/internal/processor/MethodRetrievalProcessor.java

```diff
@@ -215,12 +221,19 @@ private TypeElement asTypeElement(DeclaredType type) {
     }
 
     private SourceMethod getMethod(TypeElement usedMapper,
+                                   Type useContext,
                                    ExecutableElement method,
                                    TypeElement mapperToImplement,
                                    MapperOptions mapperOptions,
                                    List<SourceMethod> prototypeMethods) {
-
-        ExecutableType methodType = typeFactory.getMethodType( (DeclaredType) usedMapper.asType(), method );
+        DeclaredType methodDeclaredType;
+        if ( useContext != null ) {
+            methodDeclaredType = (DeclaredType) useContext.getTypeMirror();
+        }
+        else {
+            methodDeclaredType = (DeclaredType) usedMapper.asType();
+        }
```

Comment:
I am not too fond of this approach. Why don't we change the first method of the `retrieveMethods ` and instead of passing `TypeElement` for the `usedMapper` we can pass our `Type`.

Reply:
I thought about doing this, but I'm not sure if multiple types such as `Type `and `TypeElement `can be converted safely. I will try again.